packages/zoe/src/zoeService/zoe.js

```diff
@@ -193,6 +200,33 @@ const makeDurableZoeKit = ({
         },
       );
     },
+    // TODO(8686) Remove this method once empty payments have been dropped
+    async releaseEmptyPayments(issuer, brand, max) {
+      const purse = dataAccess.provideLocalPurse(issuer, brand);
+      const recoverySet = purse.getRecoverySet();
+
+      await null;
+      trace(`${brand} recoverySet has ${recoverySet.payload.length} payments`);
+      let count = 0;
+      // recoverySet claims to be a copySet, but it's a record with a payload
+      // field containing an array of payment promises
+      for (const p of recoverySet.payload) {
```

Comment:
```suggestion
      for (const p of getCopySetKeys(recoverySet)) {
```

The comment is correct, but for the "but". `recoverySet` is a CopySet. A CopySet is a Tagged, which is a record-like structure with a tag (accessed by `getTag`) and a `payload` field, as you discovered. However, as a user of a copySet, you should generally avoid assuming too much about its internal implementation. Rather, use the `getCopySetKeys` function for obtaining those keys. Its implementation at https://github.com/endojs/endo/blob/0a91fbe7a066ef3d754b350d5da7ccf55638c12d/packages/patterns/src/keys/checkKey.js#L200-L209 is

```js
export const getCopySetKeys = s => {
  assertCopySet(s);
  return s.payload;
};
```

Reply:
I looked and didn't find useful declarations or documentation. The wonderful endojs docs do link from [CopySet](https://endojs.github.io/endo/types/_endo_patterns.CopySet.html) to [CopyTagged](https://endojs.github.io/endo/types/_endo_captp.CopyTagged.html), but I didn't find anything to suggest that `getCopySetKeys()` was the appropriate way to access it. If I wasn't just looking in the wrong place, where can we add it so others will know?

Reply:
I updated the code and dropped the comment.

Reply:
> I looked and didn't find useful declarations or documentation. The wonderful endojs docs do link from [CopySet](https://endojs.github.io/endo/types/_endo_patterns.CopySet.html) to [CopyTagged](https://endojs.github.io/endo/types/_endo_captp.CopyTagged.html), but I didn't find anything to suggest that getCopySetKeys() was the appropriate way to access it. If I wasn't just looking in the wrong place, where can we add it so others will know?

Please file a bug on this?

Reply:
https://github.com/endojs/endo/issues/2299